### PR TITLE
cheribsdtest: Add tests in anticipation of c18n where function pointers are wrapped by default

### PR DESCRIPTION
**Pull request by dpgao**
Add test for global integer variables containing function addresses.
Also increases the length of signal_returncap because c18n with default-wrapped trampolines requires that.

### CODE REVIEW

File: bin/cheribsdtest/cheribsdtest_fptr_canon.c [hunk 1]
```diff
@@ -82,3 +84,20 @@ CHERIBSDTEST(fptr_canon_dlfunc,
 
 	cheribsdtest_success();
 }
+
+CHERIBSDTEST(fptr_canon_scalar,
+    "Check that function pointers are canonical when cast to scalar")
```
Comment (jrtc27) [inline bin/cheribsdtest/cheribsdtest_fptr_canon.c:89]: As with LLD, please don't use the term scalar. A capability is still a scalar (contrast with a vector of capabilities).

File: bin/cheribsdtest/cheribsdtest_signal.c [hunk 2]
```diff
@@ -267,6 +267,8 @@ CHERIBSDTEST(signal_returncap,
 	CHERIBSDTEST_VERIFY2(v == CHERI_CAP_USER_CODE_LENGTH,
 	    "length 0x%jx (expected <= 0x%jx)", v,
 	    (uintmax_t)CHERI_CAP_USER_CODE_LENGTH);
+#elif defined(CHERIBSD_C18N_TESTS)
+	CHERIBSDTEST_VERIFY2(v <= 0x300, "length 0x%jx (expected <= 0x300)", v);
```
Comment (jrtc27) [inline bin/cheribsdtest/cheribsdtest_signal.c:271]: Unrelated?

File: bin/cheribsdtest/cheribsdtest_fptr_canon.c (hunk 1 above)
Comment (jrtc27) [reply bin/cheribsdtest/cheribsdtest_fptr_canon.c:89]: Also, this isn't about casting, this is about loading from memory where the value in memory is an integer address that's been relocated, rather than a pointer.

File: lib/libcheribsdtest_dynamic/cheribsdtest_dynamic_fptr.c [hunk 3]
```diff
@@ -44,3 +44,14 @@ void
 {
 	return (&cheribsdtest_dynamic_dummy_func);
 }
+
+/*
+ * This function must only be referenced as a scalar.
+ */
+void
+cheribsdtest_dynamic_dummy_func_scalar_only(void)
```
Comment (jrtc27) [inline lib/libcheribsdtest_dynamic/cheribsdtest_dynamic_fptr.c:52]: Why do we need a new function?

Comment (dpgao) [reply lib/libcheribsdtest_dynamic/cheribsdtest_dynamic_fptr.c:52]: I’m aiming to test the ‘lazy-trampolines’ that come with default-wrapped function pointers. To do this you’d need to first resolve an integer address to the function, then resolve a capability, and make sure that a) they are equal and b) the latter is a tagged capability. (I haven’t actually tested (b) here because that would assume the existence of CHERI. To avoid using ifdefs maybe we can just call the function?)

The new function is needed to provide this ordering of symbol resolution.

File: bin/cheribsdtest/cheribsdtest_signal.c (hunk 2 above)
Comment (dpgao) [reply bin/cheribsdtest/cheribsdtest_signal.c:271]: No, rather the PR’s title is a bit misleading.

**Comment by dpgao** [comment]
On second thought, cheribsdtest is not the best place to test such an implementation detail as the lazy trampoline. I've changed the test as you suggested.

**Comment by dpgao** [comment]
> If there's an interesting corner case like that it's good to have a test for it, I'd just like to fully understand what exactly you're trying to test. It should probably be a separate test that's specifically for lazy trampolines though, so it's clear that it's testing that, rather than also implicitly being the test for R_AARCH64_FUNC_RELATIVE.

OK. Let's get this PR through as-is for now and I'll think about adding new tests in the future.

**Comment by gvnn3** [comment]
Bump.

File: lib/libcheribsdtest_dynamic/cheribsdtest_dynamic_fptr.c [hunk 4]
```diff
@@ -44,3 +44,11 @@ void
 {
 	return (&cheribsdtest_dynamic_dummy_func);
 }
+
+ptraddr_t
+cheribsdtest_dynamic_get_dummy_fptr_addr(void)
+{
+    static volatile ptraddr_t addr =
+        (ptraddr_t)&cheribsdtest_dynamic_dummy_func;
+	return (addr);
```
[the remark this answers is not in the document]
Comment (dpgao) [reply lib/libcheribsdtest_dynamic/cheribsdtest_dynamic_fptr.c]: Fixed.

File: bin/cheribsdtest/cheribsdtest_signal.c (hunk 2 above)
Comment (bsdjhb) [reply bin/cheribsdtest/cheribsdtest_signal.c:271]: I would suggest fixing up the comments here, something like:

```
     v = cheri_getlen(handler_returncap);
#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
     /* The purecap ABI does not bound PCC capabilities. */
     ...
#elif defined(CHERIBSD_C18N_TESTS)
     /* Signal handlers return to a c18n trampoline. */
     ...
#else
     /* 256 bytes should be more than enough to cover sigcode. */
    ...
#endif
```

Of course, one could argue that instead of duplicating the bulk of the code, we should just define a helper macro (e.g. `SIGNAL_RETURNCAP_EXPECTED_LEN`) with the #if sequence above used to define the macro (and the comments pulled up there) and here in the code it would just be the middle code case but with `CHERI_CAP_USER_CODE_LENGTH` replaced with `SIGNAL_RETURNCAP_EXPECTED_LEN`.

File: bin/cheribsdtest/cheribsdtest_fptr_canon.c [hunk 5]
```diff
@@ -82,3 +84,21 @@ CHERIBSDTEST(fptr_canon_dlfunc,
...
+{
+	volatile ptraddr_t fptr_int;
+	void (* volatile fptr_dlsym)(void);
+
+	fptr_int = cheribsdtest_dynamic_get_dummy_fptr_addr();
+	fptr_dlsym = (void (*)(void))dlsym(RTLD_DEFAULT,
+	    "cheribsdtest_dynamic_dummy_func");
+
+	CHERIBSDTEST_VERIFY2(fptr_int == (ptraddr_t)fptr_dlsym,
+	    "Integer address %#p differs from dlsym %#p",
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_fptr_canon.c:100]: Hmm, I would just use `%p` for `fptr_int` here.  We know it won't have any attributes?  Alternatively, perhaps print it as a hex uintmax_t instead of a pointer?
